lib/assoc/updater.ex

```diff
@@ -44,49 +44,59 @@ defmodule Assoc.Updater do
   @doc """
   Create and update associated records.
   """
-  @spec update_associations(map(), struct() | {:ok, struct()}, map()) :: {:ok, Map.t} | {:error, String.t}
-  def update_associations(repo, {:ok, parent}, params), do: update_associations(repo, parent, params)
-  def update_associations(repo, %schema{} = parent, params) do
-    associations = Enum.reduce(schema.updatable_associations, %{}, fn ({association_key, association_schema}, acc) ->
-      association_params =
-        params
-        |> Assoc.Util.keys_to_atoms()
-        |> Map.get(association_key, :omitted)
-
-      case update_association(repo, parent, association_schema, association_params) do
-        {:skipped, _} -> acc
-        {:error, error} -> Map.put(acc, association_key, error)
-        {:ok, result} -> Map.put(acc, association_key, result)
-      end
-    end)
+  @spec update_associations(map(), struct() | {:ok, struct()}, map()) ::
+          {:ok, Map.t()} | {:error, String.t()}
+  def update_associations(repo, {:ok, parent}, params),
+    do: update_associations(repo, parent, params)
+
+  def update_associations(repo, schema = %{} = parent, params) do
```

Comment:
I think the refactor to this line is the cause of the test failures.

```
- def update_associations(repo, %schema{} = parent, params) do
+ def update_associations(repo, schema = %{} = parent, params)
```

The reason is these are not equivalent:

```diff
- %schema{} = parent
+ schema = %{} = parent
```

The `%schema{} = ...` syntax is a special way to get the struct name using pattern matching: https://stackoverflow.com/questions/39618208/how-to-get-struct-name-in-elixir

The revised code returns the struct value (e.g. `%User{...}`), not the struct name (`User`).

Reply:
When I reverted the code to the original, I was able to see the warning you were trying to resolve:

```
warning: incompatible types:
    map() !~ atom()
in expression:
    # lib/assoc/updater.ex:54
    schema.updatable_associations
where "schema" was given the type atom() in:
    # lib/assoc/updater.ex:52
    %schema{}
where "schema" was given the type map() (due to calling var.field) in:
    # lib/assoc/updater.ex:54
    schema.updatable_associations
HINT: "var.field" (without parentheses) implies "var" is a map() while "var.fun()" (with parentheses) implies "var" is an atom()
Conflict found at
  lib/assoc/updater.ex:54: Assoc.Updater.update_associations/3
```

I __think__ the change I introduced here solves it: https://github.com/chrislaskey/assoc/pull/6/commits/0244662d1f3338347fa9740fc1560db05647bf97

The upshot is I stopped using the pattern matching to pull out the struct module name. And instead I went with the more explicit `__struct__`  function recommended in the StackOverflow post: https://stackoverflow.com/questions/39618208/how-to-get-struct-name-in-elixir

So references to `schema` became `parent.__struct__` :+1:

That means the original pattern matching can be modified slightly:

```diff
- %schema{} = parent
+ %_schema{} = parent
```

This makes sure the function is only executed when passed a struct. But schema is now prefixed with an underscore, since we're using `parent.__struct__` instead.

Reply:
@jmhthethird my branch should be good to pull down and validate with your codebase: https://github.com/chrislaskey/assoc/pull/6

Reply:
Thank you @chrislaskey! It is looking good on my side. Should I close this PR in favor of #6 that you would merge?

Reply:
Thanks Jack! I closed #6 in favor of this one :+1: I'll merge momentarily, cut a new tag, and push to hex.